scanner.go

```diff
@@ -359,3 +387,58 @@ func (s *scanner) closeRegionScanner() {
 	}
 	s.curRegionScannerID = noScannerID
 }
+
+// renews a scanner by resending scan request with renew = true
+func (s *scanner) renew() error {
+	select {
+	case <-s.rpc.Context().Done():
+		return errors.New("context was canceled")
+	default:
+		rpc, err := hrpc.NewScanRange(s.rpc.Context(),
+			s.rpc.Table(),
+			s.startRow,
+			nil,
+			hrpc.ScannerID(s.curRegionScannerID),
+			hrpc.NumberOfRows(s.rpc.NumberOfRows()),
+			hrpc.Priority(s.rpc.Priority()),
+			hrpc.Renew(),
+		)
+		if err != nil {
+			return err
+		}
+		res, sendErr := s.SendRPC(rpc)
+		if sendErr != nil {
+			return sendErr
+		}
+		resp, ok := res.(*pb.ScanResponse)
+		if !ok {
+			return errors.New("got non-ScanResponse for scan request")
+		}
+		// save new scanner id
+		s.curRegionScannerID = resp.GetScannerId()
+		return nil
+	}
+}
+
+// starts the renew timer, after it fires, renewScanner is called
+func (s *scanner) startRenewTimer() {
+	if s.rpc.Renew() {
```

Comment:
Added a check for whether the timer is nil at the beginning. It can be an invariant that before calling `startRenewTimer`, the `renewTimer` field must be nil. To follow that invariant, I also stopped the renew timer as well at the end of `renewScanner`.